juniper_graphql_ws/src/lib.rs

```diff
@@ -0,0 +1,1071 @@
+                                // oneshot is ever fired.
+                                let s = stream::unfold((rx, s.boxed()), |(rx, mut s)| async move {
+                                    let next = match future::select(rx, s.next()).await {
+                                        Either::Left(_) => None,
+                                        Either::Right((r, rx)) => r.map(|r| (r, rx)),
+                                    };
+                                    next.map(|(r, rx)| (r, (rx, s)))
+                                });
+
+                                // Once the stream ends, send the Complete message.
+                                let s = s.chain(
+                                    Reaction::ServerMessage(ServerMessage::Complete { id })
+                                        .to_stream(),
+                                );
+
+                                s.boxed()
+                            }
+                        }
+                    }
+                    ClientMessage::Stop { id } => {
+                        stoppers.remove(&id);
+                        stream::empty().boxed()
+                    }
+                    _ => stream::empty().boxed(),
+                };
+                (
+                    Self::Active {
+                        config,
+                        stoppers,
+                        schema,
+                    },
+                    reactions,
+                )
+            }
+            Self::Terminated => (self, stream::empty().boxed()),
+        }
+    }
+
+    async fn start(id: String, params: ExecutionParams<S>) -> BoxStream<'static, Reaction<S>> {
+        // TODO: This could be made more efficient if juniper exposed functionality to allow us to
+        // parse and validate the query, determine whether it's a subscription, and then execute
+        // it. For now, the query gets parsed and validated twice.
+
+        let params = Arc::new(params);
+
+        // Try to execute this as a query or mutation.
+        match juniper::execute(
+            &params.start_payload.query,
+            params
+                .start_payload
+                .operation_name
+                .as_ref()
+                .map(|s| s.as_str()),
+            params.schema.root_node(),
+            &params.start_payload.variables,
+            &params.config.context,
+        )
+        .await
+        {
+            Ok((data, errors)) => {
+                return Reaction::ServerMessage(ServerMessage::Data {
+                    id: id.clone(),
+                    payload: DataPayload { data, errors },
+                })
+                .to_stream();
+            }
+            Err(GraphQLError::IsSubscription) => {}
+            Err(e) => {
+                return Reaction::ServerMessage(ServerMessage::Error {
+                    id: id.clone(),
+                    payload: unsafe { ErrorPayload::new_unchecked(Box::new(params.clone()), e) },
```

Comment:
Please put a comment why this is needed and safe to do.

Reply:
Done. Side note: It might be worth adding an owned version of `GraphQLError` or giving `GraphQLError` some getters that would allow outside crates to create an owned error structure. Doing so would let me eliminate all except one `unsafe` in this PR. I was trying to be conservative with changes to the `juniper` crate here though.